File: localization-workflow/server/routes/invoices.js
const express = require('express');
const { PrismaClient } = require('@prisma/client');
const { authenticateToken, authorizeRoles, authorizeResourceAccess } = require('../middleware/auth');
const { validate, schemas } = require('../middleware/validation');

const router = express.Router();
const prisma = new PrismaClient();

// Get invoices
router.get('/', authenticateToken, async (req, res) => {
  try {
    const { status, month, year, projectId, translatorId, page = 1, limit = 10 } = req.query;
    const skip = (page - 1) * limit;

    const where = {};
    if (status) where.status = status;
    if (month) where.month = parseInt(month);
    if (year) where.year = parseInt(year);
    if (projectId) where.projectId = projectId;
    if (translatorId) where.translatorId = translatorId;

    // Filter by user role
    if (req.user.role === 'TRANSLATOR') {
      where.userId = req.user.id;
    }

    const [invoices, total] = await Promise.all([
      prisma.invoice.findMany({
        where,
        include: {
          project: {
            select: { id: true, name: true }
          },
          translator: {
            include: {
              user: {
                select: { id: true, name: true, email: true }
              }
            }
          },
          lineItems: true
        },
        skip: parseInt(skip),
        take: parseInt(limit),
        orderBy: { createdAt: 'desc' }
      }),
      prisma.invoice.count({ where })
    ]);

    res.json({
      invoices,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / limit)
      }
    });
  } catch (error) {
    console.error('Get invoices error:', error);
    res.status(500).json({ error: 'Failed to fetch invoices' });
  }
});

// Get invoice by ID
router.get('/:id', 
  authenticateToken, 
  authorizeResourceAccess('invoice'),
  async (req, res) => {
    try {
      const invoice = await prisma.invoice.findUnique({
        where: { id: req.params.id },
        include: {
          project: {
            select: { id: true, name: true, description: true }
          },
          translator: {
            include: {
              user: {
                select: { id: true, name: true, email: true }
              }
            }
          },
          lineItems: true
        }
      });

      if (!invoice) {
        return res.status(404).json({ error: 'Invoice not found' });
      }

      res.json(invoice);
    } catch (error) {
      console.error('Get invoice error:', error);
      res.status(500).json({ error: 'Failed to fetch invoice' });
    }
  }
);

// Create invoice
router.post('/', 
  authenticateToken, 
  authorizeRoles('TRANSLATOR'), 
  validate(schemas.invoice),
  async (req, res) => {
    try {
      const { projectId, month, year, lineItems } = req.body;

      // Check if translator is assigned to this project
      const projectLocale = await prisma.projectLocale.findFirst({
        where: {
          projectId,
          translatorId: req.user.translatorProfile?.id
        }
      });

      if (!projectLocale) {
        return res.status(403).json({ error: 'Not assigned to this project' });
      }

      // Calculate total amount
      const totalAmount = lineItems.reduce((sum, item) => {
        return sum + (item.wordCount * item.ratePerWord);
      }, 0);

      // Generate invoice number
      const invoiceNumber = `INV-${year}-${month.toString().padStart(2, '0')}-${Date.now()}`;

      const invoice = await prisma.invoice.create({
        data: {
          invoiceNumber,
          projectId,
          translatorId: req.user.translatorProfile.id,
          userId: req.user.id,
          month,
          year,
          totalAmount,
          lineItems: {
            create: lineItems.map(item => ({
              description: item.description,
              wordCount: item.wordCount,
              ratePerWord: item.ratePerWord,
              amount: item.wordCount * item.ratePerWord
            }))
          }
        },
        include: {
          project: {
            select: { id: true, name: true }
          },
          translator: {
            include: {
              user: {
                select: { id: true, name: true, email: true }
              }
            }
          },
          lineItems: true
        }
      });

      res.status(201).json({
        message: 'Invoice created successfully',
        invoice
      });
    } catch (error) {
      console.error('Create invoice error:', error);
      res.status(500).json({ error: 'Failed to create invoice' });
    }
  }
);

// Update invoice status
router.put('/:id/status', 
  authenticateToken,
  async (req, res) => {
    try {
      const { status } = req.body;

      const invoice = await prisma.invoice.findUnique({
        where: { id: req.params.id }
      });

      if (!invoice) {
        return res.status(404).json({ error: 'Invoice not found' });
      }

      // Check permissions
      if (req.user.role === 'TRANSLATOR' && invoice.userId !== req.user.id) {
        return res.status(403).json({ error: 'Access denied' });
      }

      // Only certain roles can approve/reject
      if (['APPROVED', 'REJECTED'].includes(status) && 
          !['ADMIN', 'PRODUCT_TEAM', 'FINANCE_TEAM'].includes(req.user.role)) {
        return res.status(403).json({ error: 'Insufficient permissions' });
      }

      // Only finance team can mark as paid
      if (status === 'PAID' && req.user.role !== 'FINANCE_TEAM') {
        return res.status(403).json({ error: 'Only finance team can mark invoices as paid' });
      }

      const updateData = { status };
      if (status === 'SUBMITTED') updateData.submittedAt = new Date();
      if (status === 'APPROVED') updateData.approvedAt = new Date();
      if (status === 'PAID') updateData.paidAt = new Date();

      const updatedInvoice = await prisma.invoice.update({
        where: { id: req.params.id },
        data: updateData,
        include: {
          project: {
            select: { id: true, name: true }
          },
          translator: {
            include: {
              user: {
                select: { id: true, name: true, email: true }
              }
            }
          },
          lineItems: true
        }
      });

      res.json({
        message: 'Invoice status updated successfully',
        invoice: updatedInvoice
      });
    } catch (error) {
      console.error('Update invoice status error:', error);
      res.status(500).json({ error: 'Failed to update invoice status' });
    }
  }
);

// Get monthly payout report
router.get('/reports/monthly-payouts', 
  authenticateToken, 
  authorizeRoles('ADMIN', 'FINANCE_TEAM'),
  async (req, res) => {
    try {
      const { month, year } = req.query;

      const where = {
        status: 'APPROVED'
      };
      if (month) where.month = parseInt(month);
      if (year) where.year = parseInt(year);

      const invoices = await prisma.invoice.findMany({
        where,
        include: {
          project: {
            select: { id: true, name: true }
          },
          translator: {
            include: {
              user: {
                select: { id: true, name: true, email: true }
              }
            }
          },
          lineItems: true
        },
        orderBy: [
          { year: 'desc' },
          { month: 'desc' },
          { translator: { user: { name: 'asc' } } }
        ]
      });

      // Group by translator
      const payoutsByTranslator = invoices.reduce((acc, invoice) => {
        const translatorId = invoice.translator.id;
        if (!acc[translatorId]) {
          acc[translatorId] = {
            translator: invoice.translator,
            totalAmount: 0,
            invoices: [],
            projects: new Set()
          };
        }
        acc[translatorId].totalAmount += invoice.totalAmount;
        acc[translatorId].invoices.push(invoice);
        acc[translatorId].projects.add(invoice.project.name);
        return acc;
      }, {});

      // Convert to array and add project names
      const payouts = Object.values(payoutsByTranslator).map(payout => ({
        ...payout,
        projects: Array.from(payout.projects)
      }));

      res.json({
        payouts,
        summary: {
          totalTranslators: payouts.length,
          totalAmount: payouts.reduce((sum, p) => sum + p.totalAmount, 0),
          totalInvoices: invoices.length
        }
      });
    } catch (error) {
      console.error('Get monthly payouts error:', error);
      res.status(500).json({ error: 'Failed to fetch monthly payouts' });
    }
  }
);

// Get spending by product report
router.get('/reports/spending-by-product', 
  authenticateToken, 
  authorizeRoles('ADMIN', 'FINANCE_TEAM', 'PRODUCT_TEAM'),
  async (req, res) => {
    try {
      const { startDate, endDate, period = 'monthly' } = req.query;

      const where = {
        status: { in: ['APPROVED', 'PAID'] }
      };

      if (startDate && endDate) {
        where.createdAt = {
          gte: new Date(startDate),
          lte: new Date(endDate)
        };
      }

      const invoices = await prisma.invoice.findMany({
        where,
        include: {
          project: {
            select: { id: true, name: true }
          }
        }
      });

      // Group by project and period
      const spendingByProduct = invoices.reduce((acc, invoice) => {
        const projectName = invoice.project.name;
        const date = new Date(invoice.createdAt);
        let periodKey;

        if (period === 'yearly') {
          periodKey = date.getFullYear().toString();
        } else if (period === 'quarterly') {
          const quarter = Math.ceil((date.getMonth() + 1) / 3);
          periodKey = `${date.getFullYear()}-Q${quarter}`;
        } else {
          periodKey = `${date.getFullYear()}-${(date.getMonth() + 1).toString().padStart(2, '0')}`;
        }

        if (!acc[projectName]) {
          acc[projectName] = {};
        }
        if (!acc[projectName][periodKey]) {
          acc[projectName][periodKey] = 0;
        }
        acc[projectName][periodKey] += invoice.totalAmount;

        return acc;
      }, {});

      res.json({ spendingByProduct });
    } catch (error) {
      console.error('Get spending by product error:', error);
      res.status(500).json({ error: 'Failed to fetch spending by product' });
    }
  }
);

module.exports = router;

File: backend/routers/invoices.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import func, extract, and_
from typing import List, Optional
from datetime import datetime, date
from database import get_db
from models import Invoice, User, UserRole, InvoiceStatus, Translator, Project, TranslationTask
from schemas import Invoice as InvoiceSchema, InvoiceCreate, InvoiceUpdate
from auth import get_current_active_user, require_finance_or_admin, require_translator_or_admin

router = APIRouter()

def generate_invoice_number(db: Session, year: int, month: int) -> str:
    """Generate a unique invoice number"""
    count = db.query(Invoice).filter(
        extract('year', Invoice.created_at) == year,
        extract('month', Invoice.created_at) == month
    ).count()
    return f"INV-{year}{month:02d}-{count + 1:04d}"

@router.post("/", response_model=InvoiceSchema)
async def create_invoice(
    invoice: InvoiceCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_translator_or_admin)
):
    """Create a new invoice"""
    # Verify translator exists
    translator = db.query(Translator).filter(Translator.id == invoice.translator_id).first()
    if not translator:
        raise HTTPException(status_code=404, detail="Translator not found")
    
    # Check permissions - translators can only create their own invoices
    if (current_user.role == UserRole.TRANSLATOR and 
        current_user.translator_profile and
        current_user.translator_profile.id != invoice.translator_id):
        raise HTTPException(status_code=403, detail="Not authorized to create invoice for this translator")
    
    # Verify project exists
    project = db.query(Project).filter(Project.id == invoice.project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Generate invoice number
    now = datetime.now()
    invoice_number = generate_invoice_number(db, now.year, now.month)
    
    db_invoice = Invoice(
        **invoice.dict(),
        invoice_number=invoice_number,
        submitted_at=datetime.utcnow()
    )
    db.add(db_invoice)
    db.commit()
    db.refresh(db_invoice)
    return db_invoice

@router.get("/", response_model=List[InvoiceSchema])
async def list_invoices(
    skip: int = 0,
    limit: int = 100,
    translator_id: Optional[int] = None,
    project_id: Optional[int] = None,
    status: Optional[InvoiceStatus] = None,
    year: Optional[int] = None,
    month: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """List invoices with filtering"""
    query = db.query(Invoice)
    
    # Apply filters
    if translator_id:
        query = query.filter(Invoice.translator_id == translator_id)
    if project_id:
        query = query.filter(Invoice.project_id == project_id)
    if status:
        query = query.filter(Invoice.status == status)
    if year:
        query = query.filter(extract('year', Invoice.created_at) == year)
    if month:
        query = query.filter(extract('month', Invoice.created_at) == month)
    
    # Role-based filtering
    if current_user.role == UserRole.TRANSLATOR and current_user.translator_profile:
        query = query.filter(Invoice.translator_id == current_user.translator_profile.id)
    elif current_user.role == UserRole.PRODUCT:
        # Product team can see invoices for their projects
        query = query.join(Project).filter(Project.created_by_id == current_user.id)
    
    invoices = query.offset(skip).limit(limit).all()
    return invoices

@router.get("/{invoice_id}", response_model=InvoiceSchema)
async def get_invoice(
    invoice_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get a specific invoice"""
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        raise HTTPException(status_code=404, detail="Invoice not found")
    
    # Check permissions
    if current_user.role == UserRole.TRANSLATOR:
        if not current_user.translator_profile or current_user.translator_profile.id != invoice.translator_id:
            raise HTTPException(status_code=403, detail="Not authorized to view this invoice")
    elif current_user.role == UserRole.PRODUCT:
        project = db.query(Project).filter(Project.id == invoice.project_id).first()
        if project.created_by_id != current_user.id:
            raise HTTPException(status_code=403, detail="Not authorized to view this invoice")
    elif current_user.role not in [UserRole.ADMIN, UserRole.FINANCE]:
        raise HTTPException(status_code=403, detail="Not authorized to view this invoice")
    
    return invoice

@router.put("/{invoice_id}", response_model=InvoiceSchema)
async def update_invoice(
    invoice_id: int,
    invoice_update: InvoiceUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Update an invoice"""
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        raise HTTPException(status_code=404, detail="Invoice not found")
    
    # Check permissions based on what's being updated
    if invoice_update.status:
        # Only finance/admin can change status
        if current_user.role not in [UserRole.ADMIN, UserRole.FINANCE]:
            raise HTTPException(status_code=403, detail="Not authorized to change invoice status")
        
        # Set timestamps based on status
        if invoice_update.status == InvoiceStatus.APPROVED:
            invoice_update.approved_at = datetime.utcnow()
        elif invoice_update.status == InvoiceStatus.PAID:
            invoice_update.paid_at = datetime.utcnow()
    
    # Translators can only update their own invoices and only certain fields
    if current_user.role == UserRole.TRANSLATOR:
        if not current_user.translator_profile or current_user.translator_profile.id != invoice.translator_id:
            raise HTTPException(status_code=403, detail="Not authorized to update this invoice")
        # Translators can only update notes if invoice is still draft
        if invoice.status != InvoiceStatus.DRAFT:
            raise HTTPException(status_code=400, detail="Cannot update submitted invoice")
    
    # Update fields
    update_data = invoice_update.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(invoice, field, value)
    
    db.commit()
    db.refresh(invoice)
    return invoice

@router.post("/{invoice_id}/approve")
async def approve_invoice(
    invoice_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_finance_or_admin)
):
    """Approve an invoice (Finance/Admin only)"""
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        raise HTTPException(status_code=404, detail="Invoice not found")
    
    if invoice.status != InvoiceStatus.SUBMITTED:
        raise HTTPException(status_code=400, detail="Only submitted invoices can be approved")
    
    invoice.status = InvoiceStatus.APPROVED
    invoice.approved_at = datetime.utcnow()
    
    db.commit()
    db.refresh(invoice)
    
    return {"message": "Invoice approved successfully", "invoice_id": invoice_id}

@router.post("/{invoice_id}/reject")
async def reject_invoice(
    invoice_id: int,
    reason: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_finance_or_admin)
):
    """Reject an invoice (Finance/Admin only)"""
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        raise HTTPException(status_code=404, detail="Invoice not found")
    
    if invoice.status not in [InvoiceStatus.SUBMITTED, InvoiceStatus.APPROVED]:
        raise HTTPException(status_code=400, detail="Cannot reject invoice in current status")
    
    invoice.status = InvoiceStatus.REJECTED
    invoice.notes = f"{invoice.notes}\n\nRejection reason: {reason}" if invoice.notes else f"Rejection reason: {reason}"
    
    db.commit()
    db.refresh(invoice)
    
    return {"message": "Invoice rejected", "invoice_id": invoice_id, "reason": reason}

@router.post("/{invoice_id}/mark-paid")
async def mark_invoice_paid(
    invoice_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_finance_or_admin)
):
    """Mark an invoice as paid (Finance/Admin only)"""
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        raise HTTPException(status_code=404, detail="Invoice not found")
    
    if invoice.status != InvoiceStatus.APPROVED:
        raise HTTPException(status_code=400, detail="Only approved invoices can be marked as paid")
    
    invoice.status = InvoiceStatus.PAID
    invoice.paid_at = datetime.utcnow()
    
    db.commit()
    db.refresh(invoice)
    
    return {"message": "Invoice marked as paid", "invoice_id": invoice_id}

@router.post("/generate-monthly")
async def generate_monthly_invoice(
    translator_id: int,
    project_id: int,
    year: int,
    month: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_translator_or_admin)
):
    """Generate monthly invoice for a translator's work on a project"""
    # Verify translator exists
    translator = db.query(Translator).filter(Translator.id == translator_id).first()
    if not translator:
        raise HTTPException(status_code=404, detail="Translator not found")
    
    # Check permissions
    if (current_user.role == UserRole.TRANSLATOR and 
        current_user.translator_profile and
        current_user.translator_profile.id != translator_id):
        raise HTTPException(status_code=403, detail="Not authorized to generate invoice for this translator")
    
    # Get completed tasks for the period
    from datetime import date
    period_start = date(year, month, 1)
    if month == 12:
        period_end = date(year + 1, 1, 1)
    else:
        period_end = date(year, month + 1, 1)
    
    tasks = db.query(TranslationTask).filter(
        and_(
            TranslationTask.translator_id == translator_id,
            TranslationTask.project_id == project_id,
            TranslationTask.status == TaskStatus.COMPLETED,
            TranslationTask.completed_at >= period_start,
            TranslationTask.completed_at < period_end
        )
    ).all()
    
    if not tasks:
        raise HTTPException(status_code=404, detail="No completed tasks found for the period")
    
    # Calculate line items and total
    line_items = []
    total_amount = 0
    
    for task in tasks:
        amount = task.actual_cost or task.estimated_cost
        line_items.append({
            "task_id": task.id,
            "description": f"Translation for {task.locale} - {task.word_count} words",
            "quantity": task.word_count,
            "rate": amount / task.word_count if task.word_count > 0 else 0,
            "amount": amount
        })
        total_amount += amount
    
    # Generate invoice number
    invoice_number = generate_invoice_number(db, year, month)
    
    # Create invoice
    db_invoice = Invoice(
        invoice_number=invoice_number,
        translator_id=translator_id,
        project_id=project_id,
        period_start=period_start,
        period_end=period_end,
        total_amount=total_amount,
        line_items=line_items,
        status=InvoiceStatus.DRAFT
    )
    
    db.add(db_invoice)
    db.commit()
    db.refresh(db_invoice)
    
    return {
        "message": "Monthly invoice generated successfully",
        "invoice": db_invoice,
        "tasks_included": len(tasks)
    }

@router.get("/monthly-payouts/{year}/{month}")
async def get_monthly_payouts(
    year: int,
    month: int,
    translator_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_finance_or_admin)
):
    """Get monthly payout report (Finance/Admin only)"""
    query = db.query(Invoice).filter(
        extract('year', Invoice.period_start) == year,
        extract('month', Invoice.period_start) == month,
        Invoice.status.in_([InvoiceStatus.APPROVED, InvoiceStatus.PAID])
    )
    
    if translator_id:
        query = query.filter(Invoice.translator_id == translator_id)
    
    invoices = query.all()
    
    # Group by translator
    payouts_by_translator = {}
    total_payout = 0
    
    for invoice in invoices:
        translator_name = invoice.translator.user.full_name
        project_name = invoice.project.name
        
        if translator_name not in payouts_by_translator:
            payouts_by_translator[translator_name] = {
                "translator_id": invoice.translator_id,
                "total_amount": 0,
                "projects": [],
                "invoices": []
            }
        
        payouts_by_translator[translator_name]["total_amount"] += invoice.total_amount
        payouts_by_translator[translator_name]["projects"].append(project_name)
        payouts_by_translator[translator_name]["invoices"].append({
            "invoice_id": invoice.id,
            "invoice_number": invoice.invoice_number,
            "amount": invoice.total_amount,
            "status": invoice.status,
            "project": project_name
        })
        total_payout += invoice.total_amount
    
    # Remove duplicate projects
    for translator_data in payouts_by_translator.values():
        translator_data["projects"] = list(set(translator_data["projects"]))
    
    return {
        "year": year,
        "month": month,
        "total_payout": total_payout,
        "payouts_by_translator": payouts_by_translator,
        "total_translators": len(payouts_by_translator)
    }

@router.get("/spending-by-product/{year}")
async def get_spending_by_product(
    year: int,
    month: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_finance_or_admin)
):
    """Get spending breakdown by product/project"""
    query = db.query(Invoice).filter(
        extract('year', Invoice.period_start) == year,
        Invoice.status == InvoiceStatus.PAID
    )
    
    if month:
        query = query.filter(extract('month', Invoice.period_start) == month)
    
    invoices = query.all()
    
    spending_by_project = {}
    total_spending = 0
    
    for invoice in invoices:
        project_name = invoice.project.name
        if project_name not in spending_by_project:
            spending_by_project[project_name] = {
                "project_id": invoice.project_id,
                "total_amount": 0,
                "invoice_count": 0
            }
        
        spending_by_project[project_name]["total_amount"] += invoice.total_amount
        spending_by_project[project_name]["invoice_count"] += 1
        total_spending += invoice.total_amount
    
    return {
        "year": year,
        "month": month,
        "total_spending": total_spending,
        "spending_by_project": spending_by_project
    }